### ISSUE TITLE
Convenience methods for Pillow integration

### ISSUE THREAD

**Issue by jonashaag** [libvips/pyvips#486]
Proposal: Add convenience integration with PIL/Pillow, eg. converting back and forth. Something along the lines of

```
pyvips.Image.from_pil(pil_img)
```

and

```
vips_img.pil()  # -> PIL.Image.Image
```

Happy to contribute a first draft if this is accepted.

**Comment by jcupitt** [comment]
Hi @jonashaag, sure this sounds useful. What would it add over going via numpy?

**Comment by jonashaag** [comment]
You're right that it's already quite simple... my suggestion of `from_pil` is the same as `new_from_array` (it's just a bit confusingly named), and my convenience methods for the other way would be

```py
    def pil(self):
        try:
            from PIL import Image
        except ModuleNotFoundError as e:
            raise ModuleNotFoundError("PIL library must be installed") from e

        return Image.fromarray(self.numpy())
```

Honestly not sure anymore if it's worth adding this because it's a _very_ thin wrapper...

**Comment by jonashaag** [comment]
@jcupitt just stumbled across an issue with this method: If the image is 16 bit, Pillow can't read it (no support). Should we make the `numpy()` method output 8 bit? Or should we add the convenience method for this case?

**Comment by jcupitt** [comment]
Hi again @jonashaag,

I guess you'd need:

```python
pil_image = PIL.Image.fromarray(image.cast("uchar", shift=True).numpy())
```

ie. shift the top 8 bits down. 

Now I think about it, there are potentially quite a few conversions you might want to do. For example, libvips images can have any number of colour channels, and any number of alpha channels. Pixels can be float or double, and they can be complex numbers. They can also be in colourspaces like CIELAB and scRGB.

We could try to make something to automatically convert all these formats, but it probably would not be correct for everyone. Maybe this is better left to library users?

**Comment by jonashaag** [comment]
IMO we should do a best effort conversion for the users who just want smooth interoperability.

If we don't, people will end up rolling their own. (I guess nowadays LLMs will do it, and that code will be mostly correct?)

**Comment by jcupitt** [comment]
Ah that's true.

Looking the PIL modes, there are more than I remembered:

https://pillow.readthedocs.io/en/stable/handbook/concepts.html#modes

Picking the best mode for an arbitrary libvips image will not be easy! Maybe something simple that just covers the common cases would be enough for now -- it can always be extended. A PR sounds great. Animated images would be interesting to support too I guess.

There's slightly similar code in vipsdisp:

https://github.com/libvips/vipsdisp/blob/master/src/tilesource.c#L594-L762

That thing takes any vips image and tries to make an 8-bit RGB or RGBA image that can be sent to the display. Maybe there are some ideas there you could borrow.

**Comment by jonashaag** [comment]
Great, I'll see what I can do.

Do you want me to modify/extend `numpy()` or add a separate method?

**Comment by jcupitt** [comment]
People the `.numpy()` method for non-PIL images, so I think that needs to stay. 

Maybe a new `pyvips.Image.pil()` method that does `.numpy().numpy_to_pil()`, perhaps also using some information from the original libvips image, such as animation state.
